reconcile/external_resources/manager.py

```diff
@@ -0,0 +1,578 @@
+            metrics.set_gauge(
+                ExternalResourcesReconcileErrorsGauge(
+                    provision_provider=r.key.provision_provider,
+                    provisioner_name=r.key.provisioner_name,
+                    provider=r.key.provider,
+                    identifier=r.key.identifier,
+                ),
+                float(state.reconciliation_errors),
+            )
+            self._update_in_progress_state(r, state)
+            if self.resource_needs_reconciliation(
+                reconciliation=r, module=module, state=state
+            ):
+                self.impl.reconcile_resource(
+                    reconciliation=r, module=module, dry_run=self.dry_run
+                )
+                self._update_state(r, state)
+
+    def _wait_for_running_jobs(
+        self,
+        jobs: Iterable[Reconciliation],
+        wait_before_start: bool = True,
+    ) -> Tuple[set[Reconciliation], set[Reconciliation]]:
+        running_jobs = set(jobs)
+        success_jobs: set[Reconciliation] = set()
+        error_jobs: set[Reconciliation] = set()
+
+        if wait_before_start and len(running_jobs) > 0:
+            logging.info("Waiting for jobs population")
+            time.sleep(5)
+
+            while running_jobs:
+                logging.info(running_jobs)
+                for reconciliation in list(running_jobs):
+                    status = self.impl.get_resource_reconcile_status(
+                        reconciliation, from_inventory=False
+                    )
+                    logging.info(status)
+                    match status:
+                        case ReconcileStatus.SUCCESS:
+                            success_jobs.add(reconciliation)
+                            running_jobs.remove(reconciliation)
+                        case ReconcileStatus.ERROR:
+                            error_jobs.add(reconciliation)
+                            running_jobs.remove(reconciliation)
+                if running_jobs:
+                    logging.info("Waiting for jobs to complete.")
+                    time.sleep(30)
+        return (success_jobs, error_jobs)
+
+    def handle_dry_run_resources(self, specs: Iterable[ExternalResourceSpec]) -> None:
+        desired_r = self._get_desired_objects_reconciliations(specs)
+        deleted_r = self._get_deleted_objects_reconciliations(specs)
+        running_jobs = set[Reconciliation]()
+        for r in desired_r.union(deleted_r):
+            m_key = ExternalResourceModuleKey(
+                provision_provider=r.key.provision_provider, provider=r.key.provider
+            )
+            module = self.modules[m_key]
+            self.impl.reconcile_resource(
+                reconciliation=r,
+                module=module,
+                dry_run=self.dry_run,
+            )
+            running_jobs.add(r)
+
+        success_jobs, error_jobs = self._wait_for_running_jobs(running_jobs)
```

Comment:
Why is dry-run handling job results? Wouldn't that block MRs?

Reply:
MRs need to report the status of the JOBS (TF PLAN) back to the PR to feed the information back to users. Each resource change runs in a JOB in a dry-run mode (Tf plan) and gets the results.